lib/confluence/api_velocity.es6.js

```diff
@@ -31,90 +29,73 @@ foam.CLASS({
-      },
-    },
-    {
-      name: 'groupReleases',
-      documentation: `Produce a nested GROUP_BY sink that groups releases by
-          browser name, then by OS name. Releases are ordered by release date.`,
-      code: function() {
-        this.info('API velocity grouping releases');
-        return this.releaseDAO.orderBy(this.Release.RELEASE_DATE).select(
-          this.GROUP_BY(this.Release.BROWSER_NAME, this.GROUP_BY(
-              this.Release.OS_NAME, this.ArraySink.create())));
-      },
+      name: 'recontextualize',
+      documentation: `Recreate object in this computer's context. This is often
+          necessary for join queries that may be based on an object that came
+          from a foreign context (and therefore may have been instantiated in a
+          whitelisted deserialization context).`,
+      code: function(o) { return o.cls_.create(o, this); },
     },
     {
       name: 'compute',
       documentation: `Compute metrics for a group of releases from a particular
           browser/platform.`,
-      code: function(releases) {
-        if (releases.length === 0) return undefined;
-        this.metricLogger.startCompute(releases[0].releaseDate, releases[0]);
-        var promises = [this.computeForSingleRelease_(releases[0])];
-        for (var i = 0; i < releases.length - 1; i++) {
-          promises.push(this.computeForPair_(releases[i], releases[i + 1]));
-        }
-        return Promise.all(promises);
+      code: function(releases, date) {
+        const numReleases = releases.length;
+        foam.assert(numReleases === 1 || numReleases === 2,
+                    'ApiVelocity.compute(): Expected 1 or 2 releases');
+        const lastReleaseDate = releases[releases.length -1].releaseDate;
+        foam.assert(
+            date.getTime() === lastReleaseDate.getTime(),
+            `ApiVelocity.compute(): Expected date to match last release date:
+                Date: ${date}
+                Last release date: ${lastReleaseDate}`);
+
+        if (numReleases === 1)
+          return this.computeForSingleRelease_(releases[0]);
+        else
+          return this.computeForPair_(releases[0], releases[1]);
       },
     },
     {
       name: 'computeForSingleRelease_',
       documentation: `Compute single metric value for first release in a
           sequence.`,
       code: function(release) {
-        this.metricLogger.startComputeForReleases(
-            release.releaseDate, release, [], []);
-        return release.interfaces.dao.select(this.COUNT())
-            .then(count => this.apiVelocityDAO.put(this.ApiVelocityData.create({
-              releaseDate: release.releaseDate,
-              browserName: release.browserName,
-              currRelease: release,
-              totalApis: count.value,
-            })));
+        this.info(`Computing ApiVelocity for single release ${release.id}`);
+        return this.recontextualize(release).interfaces.dao.select(this.COUNT())
```

Comment:
recontextualize feels like a thing that should be done by the user of this object. If I'm following things correctly, this is being done because the service might have been instantiated with a different context than the code calling the service. Why isn't something in the service layer dealing with this? Putting it there means that this code can be used without the service wrapper in a way that doesn't copy every input object.

Reply:
Done. Also fixed contextualization bug in `*_json_dao` that was causing some instances of "recontextualization needed".